### PR TITLE
Validate autoupdate_config and autoupdate_agent_rollout

### PR DESCRIPTION
**Pull request by hugoShaka**
Part of: [RFD-184](https://github.com/gravitational/teleport/pull/47126)

Goal (internal): https://github.com/gravitational/cloud/issues/10289

This PR removes the restrictions of the `autoupdate_agent_rollout` and `autoupdate_config` schedules but adds groups validation.

It also adds some optional server-side validation that should not be enforced at the resource level.

### CODE REVIEW

File: lib/auth/autoupdate/autoupdatev1/service.go [hunk 1]
```diff
@@ -589,3 +608,78 @@ func (s *Service) emitEvent(ctx context.Context, e apievents.AuditEvent) {
...
+			}
+
+			if !maps.Equal(cloudWeekdays, weekdays) {
+				return trace.BadParameter("weekdays must be set to %v in cloud", cloudGroupUpdateDays)
+			}
+
+			waitDays += int(group.WaitDays)
+		}
+		if waitDays > maxWaitDaysCLoud {
+			return trace.BadParameter("cloud support ap most %d wait days, got %d", maxWaitDaysCLoud, waitDays)
```
Comment (sclevine) [inline lib/auth/autoupdate/autoupdatev1/service.go:680]:
```suggestion
			return trace.BadParameter("cloud supports at most %d wait days, got %d", maxWaitDaysCLoud, waitDays)
```

File: lib/auth/autoupdate/autoupdatev1/service.go [hunk 2]
```diff
@@ -589,3 +608,78 @@ func (s *Service) emitEvent(ctx context.Context, e apievents.AuditEvent) {
...
+func (s *Service) validateServerSideAgentConfig(config *autoupdate.AutoUpdateConfig) error {
+	// We must check resource integrity before, because it makes no sense to try to enforce rules on an invalid resource.
+	// The generic backend service will likely check integrity again, but it's not a large performance problem.
+	err := update.ValidateAutoUpdateConfig(config)
+	if err != nil {
+		return trace.Wrap(err, "validating autoupdate config")
+	}
+
+	var maxGroups int
+	isCloud := modules.GetModules().Features().Cloud
```
Comment (sclevine) [inline lib/auth/autoupdate/autoupdatev1/service.go:641]: We may want to relax these restrictions for some Cloud customers on a case-by-case basis. Would it be tricky to add an "unrestricted" feature? (Could be a separate PR)

Comment (hugoShaka) [reply lib/auth/autoupdate/autoupdatev1/service.go:641]: We will likely want to rely on entitlements. I'm not familiar with this mechanism and will have to ask cloud devs how to do this. If it's not easy, I'll just glue an ugly environment variable.

**Review by sclevine — APPROVED** [review]

**Review by vapopov — APPROVED** [review]

**Comment by hugoShaka** [comment]
@sclevine and @vapopov  I revamped the PR as we changed from wait_days to wait_hours. Could yo do a fresh review?

File: lib/auth/autoupdate/autoupdatev1/service.go [hunk 3]
```diff
@@ -617,3 +635,126 @@ func checkAdminCloudAccess(authCtx *authz.Context) error {
...
+		maxGroups = maxGroupsTimeBasedStrategy
+	default:
+		return trace.BadParameter("unknown max group for strategy %v", agentsSpec.GetStrategy())
+	}
+
+	if len(agentsSpec.GetSchedules().GetRegular()) > maxGroups {
+		return trace.BadParameter("max groups (%d) exceeded for strategy %s, %s schedule contains %d groups", maxGroups, agentsSpec.GetStrategy(), update.AgentsScheduleRegular, len(agentsSpec.GetSchedules().GetRegular()))
+	}
+
+	if isCloud {
```
Comment (sclevine) [inline lib/auth/autoupdate/autoupdatev1/service.go:686]:
```suggestion
	if !isCloud {
		return nil
	}
```
nit: cleans up indent a bit

**Review by sclevine — APPROVED** [review]

File: lib/autoupdate/rollout/reconciler.go [hunk 4]
```diff
@@ -47,7 +47,8 @@ const (
 )
 
 var (
-	defaultUpdateDays = []string{"Mon", "Tue", "Wed", "Thu"}
+	// DefaultUpdateDays is the default list of days when groups can be updated.
+	DefaultUpdateDays = []string{"Mon", "Tue", "Wed", "Thu"}
```
Comment (vapopov) [inline lib/autoupdate/rollout/reconciler.go:51]: hope its not going to be modified externally

**Review by vapopov — APPROVED** [review]
LGTM, seems like you need to adjust unit tests (nil struct in map)

**Review by r0mant — APPROVED** [review]
Bot.
